### PR TITLE
Provider logic + Controllers

### PR DESCRIPTION
**Pull request by ignaciosantise**
## Summary
- Added provider initialization
- Added controllers based on [web3modal](https://github.com/WalletConnect/web3modal) implementation
- Added basic view to show user address + disconnect button

### iOS

https://user-images.githubusercontent.com/25931366/229607879-c122469b-f58c-403a-bab5-cc08674c7b3d.MP4

### Android

https://user-images.githubusercontent.com/25931366/231514967-62fcc8ec-f139-435e-bf5b-22e2150b72ef.mp4

### CODE REVIEW

File: example/package.json [hunk 1]
```diff
@@ -9,20 +9,22 @@
     "web": "expo start --web"
```
Comment (0xAsimetriq) [inline example/package.json:9]: Mixture of versions here, can we remove `~` and `^` in order to stick with stricter packages. Might need to remove lockfile and node_modules to re-install fresh.

File: package.json [hunk 2]
```diff
@@ -52,25 +52,26 @@
   "devDependencies": {
```
Comment (0xAsimetriq) [inline package.json:52]: Same comment as above about stricter packages.

File: src/components/Web3Button.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import React, { useEffect } from 'react';
```
Comment (0xAsimetriq) [inline src/components/Web3Button.tsx:1]: mixed use of direct imports for `useEffects` and indirect `React.useState`, should stick to one pattern.

File: src/components/Web3Modal.tsx [hunk 4]
```diff
@@ -1,127 +1,174 @@
...
-    ExplorerUtil.fetchWallets(projectId).then((wallets) => {
-      setWalletListLoading(false);
-      if (wallets) {
-        setInitialWallets(wallets.listings.slice(0, 7));
-        setAllWallets(wallets.listings);
-      }
-    });
-  }, [projectId]);
+  const onSessionError = useCallback(async () => {
+    // TODO: Improve this, check why is alerting a lot, and check MaxListeners warning
```
Comment (0xAsimetriq) [inline src/components/Web3Modal.tsx:45]: For context, `MaxListeners` is likely coming from sign sdk (we saw this there as well), unless it's something else.

File: src/components/Web3Modal.tsx [hunk 5]
```diff
@@ -1,127 +1,174 @@
...
+    const session = ClientCtrl.session();
+    if (topic === session?.topic) {
+      OptionsCtrl.resetAccount();
+      ClientCtrl.clearSession();
+    }
+  }, []);
+
+  const onConnect = useCallback(async () => {
+    const provider = ClientCtrl.provider();
+    createSession(provider)
```
Comment (0xAsimetriq) [inline src/components/Web3Modal.tsx:60]: Maybe use `async/await` with try/catch

File: src/controllers/ClientCtrl.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import { ethers } from 'ethers';
+import { proxy, ref } from 'valtio/vanilla';
```
Comment (0xAsimetriq) [inline src/controllers/ClientCtrl.ts:2]: Neat, didn't know `ref` existed 👍 will be useful in the future

**Review by 0xAsimetriq — COMMENTED** [review]
Really cool stuff!

* I left few comments on code style, that apply all around. I think it might be time to introduce tools like eslint  / prettier to catch them? Also perhaps set up github actions to run these checks on pr's (refference in web repo).

* I noticed that old explorer api is in use, we now have web3modal specific endpoints, feel free to ping me about these to discuss.

* We can probably remove x1 images (maybe even x2), as likelihood of someone using these even on older phones is very very low and we might benefit from less maintenance, but will leave this up to you.

* I usually don't leave any comments / todo's in, but since we are early in dev, it's probably ok, just make sure to not forget about them.

**Comment by ignaciosantise** [comment]
Thanks for your feedback @0xAsimetriq 🎉 

About image sizes, im following react-native [docs](https://reactnative.dev/docs/images#static-image-resources) recommendations, and i think there's no maintenance problem because Figma has the option to export in all sizes.

**Comment by crypblizz8** [comment]
> I left few comments on code style, that apply all around. I think it might be time to introduce tools like eslint / prettier to catch them? Also perhaps set up github actions to run these checks on pr's (refference in web repo).

+1 @0xAsimetriq so let's add some GHub Actions for this. Maybe also let's consider leaving an Expo Snack as our version of Vercel. Then we can easily scan without having to do the xCode / RN CLI setup each time? Do you know of any other streamlined flows @ignaciosantise ?

**Comment by crypblizz8** [comment]
NonBlocking: Let's also add a README to the main directory and for examples.

File: src/components/QRCode.tsx [hunk 7]
```diff
@@ -1,4 +1,4 @@
-import React, { useMemo } from 'react';
```
Comment (crypblizz8) [inline src/components/QRCode.tsx:1]: I sometimes found that it was necessary during RN Development to keep the `React` part but not sure if it errors. If no errors, looks good.

**Comment by crypblizz8** [comment]
IMG_8664
Running into this while doing examples. I did `yarn` and normal Expo scan/ login. Added projectID. Anything I am missing?

File: src/types/controllerTypes.ts [hunk 8]
```diff
@@ -1,12 +1,46 @@
...
+  initialized: boolean;
+  provider?: UniversalProvider;
+  web3Provider?: ethers.providers.Web3Provider;
+  session?: SessionTypes.Struct;
+}
+
+// -- ConfigCtrl ------------------------------------------- //
+export interface ConfigCtrlState {
+  projectId: string;
+  walletConnectVersion?: 1 | 2;
```
Comment (crypblizz8) [inline src/types/controllerTypes.ts:16]: Is V1 necessary? As we will not support it?

File: src/types/controllerTypes.ts [hunk 9]
```diff
@@ -1,12 +1,46 @@
...
+export interface ModalCtrlState {
+  open: boolean;
+}
+
+// -- OptionsCtrl --------------------------------------- //
+export interface OptionsCtrlState {
+  address?: string;
+  isConnected: boolean;
+  isDataLoaded: boolean;
+  walletConnectVersion: 1 | 2;
```
Comment (crypblizz8) [inline src/types/controllerTypes.ts:30]: Same here.

File: src/hooks/useWeb3Modal.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import { useEffect, useState } from 'react';
+import { ModalCtrl } from '../controllers/ModalCtrl';
+import { ClientCtrl } from '../controllers/ClientCtrl';
+
+export function useWeb3Modal() {
```
Comment (crypblizz8) [inline src/hooks/useWeb3Modal.ts:5]: I see this export + from initial search in VSCode it shows it is exported in index.tsx. Is there specific usage of this used in examples or elsewhere? Apologies if I am missing something

File: src/controllers/OptionsCtrl.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+  setIsConnected(isConnected: OptionsCtrlState['isConnected']) {
+    state.isConnected = isConnected;
+  },
+
+  setIsDataLoaded(isDataLoaded: OptionsCtrlState['isDataLoaded']) {
+    state.isDataLoaded = isDataLoaded;
+  },
+
+  setWalletConnectVersion(
```
Comment (crypblizz8) [inline src/controllers/OptionsCtrl.ts:44]: Also related to WC versioning comments made below.

File: src/constants/WalletSchemes.json
Comment (crypblizz8) [inline src/constants/WalletSchemes.json:1]: Curious/NonBlocking: Why was this deleted? was it used somewhere else?

File: src/components/QRCode.tsx (hunk 7 above)
Comment (ignaciosantise) [reply src/components/QRCode.tsx:1]: since React 17, there's no need to import React. I disabled the rule that marked the missing import as an error.

File: src/types/controllerTypes.ts (hunk 8 above)
Comment (ignaciosantise) [reply src/types/controllerTypes.ts:16]: nope, i'll remove it 👍

Comment (ignaciosantise) [reply src/types/controllerTypes.ts:16]: in fact, i could remove the whole `walletConnectVersion` version

File: src/hooks/useWeb3Modal.ts (hunk 10 above)
Comment (ignaciosantise) [reply src/hooks/useWeb3Modal.ts:5]: the exports in `index.tsx` are the "public" exports from the SDK, that are used in the example app.

`import { Web3Modal, Web3Button, useWeb3Modal } from '@web3modal/react-native';`

**Comment by ignaciosantise** [comment]
> NonBlocking: when I did a yarn on the `examples` repo, it asked me to `✔ It looks like you're trying to use TypeScript but don't have the required dependencies installed. Would you like to install typescript@^4.9.4, @types/react@~18.0.27? … yes`
> 
> Is this something we need to add or is it more an expo thing?

@crypblizz8 in order to run the example go to the root folder and:

1. Run `yarn` to install sdk and example dependencies
2. Run `yarn example ios` or `yarn example android`

**Comment by ignaciosantise** [comment]
> IMG_8664 Running into this while doing examples. I did `yarn` and normal Expo scan/ login. Added projectID. Anything I am missing?

🤔 could you run `yarn jest --clearCache` in root folder? and try again

**Comment by ignaciosantise** [comment]
> NonBlocking: when I did a yarn on the `examples` repo, it asked me to `✔ It looks like you're trying to use TypeScript but don't have the required dependencies installed. Would you like to install typescript@^4.9.4, @types/react@~18.0.27? … yes`
> 
> Is this something we need to add or is it more an expo thing?

To run this example go to the root folder and:
1. Run `yarn` to install all deps
2. Run `yarn example android` or `yarn example ios`

File: src/components/Web3Modal.tsx [hunk 12]
```diff
@@ -1,127 +1,167 @@
...
-  );
+  const subscribeToEvents = useCallback(async () => {
+    const provider = ClientCtrl.provider();
+    if (provider) {
+      provider.on('display_uri', (uri: string) => {
+        OptionsCtrl.setSessionUri(uri);
+      });
+
+      // Subscribe to session ping
+      provider.on(
```
Comment (0xAsimetriq) [inline src/components/Web3Modal.tsx:77]: Should these be unsubscribed from somewhere i.e. instead of useCallback, perhaps put them in useEffect and remove when unmounting?

File: src/views/Error.tsx [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import { View, Text } from 'react-native';
+import type { RouterProps } from '../types/routerTypes';
+
+export function Error(_: RouterProps) {
```
Comment (0xAsimetriq) [inline src/views/Error.tsx:4]: Is `_: RouterProps` bit needed given that it is not used, or is this a placeholder for future?

File: src/views/ViewAllExplorer.tsx [hunk 14]
```diff
@@ -39,27 +37,33 @@ function ViewAllExplorer({
...
           <FlatList
-            data={explorerData || []}
+            data={wallets || []}
             style={styles.list}
             contentContainerStyle={styles.listContentContainer}
             indicatorStyle={isDarkMode ? 'white' : 'black'}
             showsVerticalScrollIndicator
             numColumns={4}
-            getItemLayout={(_, index) => ({
+            getItemLayout={(__, index) => ({
```
Comment (0xAsimetriq) [inline src/views/ViewAllExplorer.tsx:57]: `__` looks weird, maybe we can have a more descriptive unused variable i.e `_something`

**Review by 0xAsimetriq — APPROVED** [review]
Looks good, left few comments 👍

File: src/components/Web3Modal.tsx (hunk 12 above)
Comment (ignaciosantise) [reply src/components/Web3Modal.tsx]: you are right, i didn't know there was a function to remove those listeners. Should we add something in the [docs](https://docs.walletconnect.com/2.0/javascript/providers/universal#events)?
